### PR TITLE
Add sarm

### PR DESCRIPTION
**Pull request by pkooij**
### **Differences compared to original:**
- Handling insufficient history (sampling):
    **LeRobot:**
    Clamps indices to valid bounds and applies copy-padding at the beginning and end.
    Samples frames in a bidirectional manner (both backward- and forward-looking).
    **Original:**
    Uses an adaptive stride when looking backward and there is not enough history.

### **Progress visualization** 
Of trained SARM model for sparse and dense annotations (sparse every frame inference, dense has stride = 30)
image 
Screenshot 2025-12-13 at 17 44 28

Inference example on failed episode,  (sparse and dense) (https://huggingface.co/datasets/lerobot-data-collection/eval_pi0_fold_11-30_1) (not in training data)
image

image

TODO:
- Test RA_BC on real robot
- Add RA_BC training loss curves

### CODE REVIEW

File: docs/source/sarm.mdx [hunk 1]
```diff
@@ -0,0 +1,617 @@
...
+  --visualize-type dense \
+  --num-visualizations 5 \
+  --video-key observation.images.base \
+  --output-dir ./subtask_viz
+```
+
+</hfoption>
+<hfoption id="dual">
+
+Visualize annotations using the `--visualize-only` flag:
```
Comment (s1lent4gnt) [inline docs/source/sarm.mdx:229]: We can remove this example,or we say this command to show `both` annotation because it looks similar to the above

File: docs/source/sarm.mdx [hunk 2]
```diff
@@ -0,0 +1,617 @@
...
+
+- **Progress plot**: Predicted progress (and optional annotation-derived “GT” when available and `--stride 1`)
+- **Stage probabilities**: Stacked area plot of predicted stage probabilities
+- **Sample frames**: Key frames from the episode with progress/stage labels
+
+### Visualization Arguments
+
+| Argument               | Description                                              |
+| ---------------------- | -------------------------------------------------------- |
+| `--visualize-only`     | Only visualize predictions (no RABC computation)         |
```
Comment (s1lent4gnt) [inline docs/source/sarm.mdx]: Change `RABC` to `RA-BC` for consistency

File: src/lerobot/configs/train.py [hunk 3]
```diff
@@ -64,9 +64,17 @@ class TrainPipelineConfig(HubMixin):
     scheduler: LRSchedulerConfig | None = None
     eval: EvalConfig = field(default_factory=EvalConfig)
     wandb: WandBConfig = field(default_factory=WandBConfig)
-    checkpoint_path: Path | None = field(init=False, default=None)
+
+    # RA-BC (Reward-Aligned Behavior Cloning) parameters
+    use_rabc: bool = False  # Enable reward-weighted training
+    rabc_progress_path: str | None = None  # Path to precomputed SARM progress parquet file
```
Comment (s1lent4gnt) [inline src/lerobot/configs/train.py:71]: Is it possible to infer the SARM progress from the dataset?

File: pyproject.toml [hunk 4]
```diff
@@ -133,6 +133,7 @@ groot = [
     "ninja>=1.11.1,<2.0.0",
     "flash-attn>=2.5.9,<3.0.0 ; sys_platform != 'darwin'"
 ]
+sarm = ["lerobot[transformers-dep]", "faker>=33.0.0,<35.0.0"]
```
Comment (s1lent4gnt) [inline pyproject.toml]: I think we need to add `matplotlib` for `subtask_annotation.py`

File: src/lerobot/data_processing/sarm_annotations/subtask_annotation.py [hunk 5]
```diff
@@ -0,0 +1,1221 @@
+#!/usr/bin/env python
```
Comment (s1lent4gnt) [inline src/lerobot/data_processing/sarm_annotations/subtask_annotation.py:1]: I got memory/RAM issue running this script with `Qwen/Qwen3-VL-4B-Instruct`. We need to investigate why?

File: src/lerobot/data_processing/sarm_annotations/subtask_annotation.py [hunk 6]
```diff
@@ -0,0 +1,1221 @@
...
+import cv2
+import matplotlib.patches as mpatches
+import matplotlib.pyplot as plt
+import numpy as np
+import pandas as pd
+import torch
+from pydantic import BaseModel, Field
+from qwen_vl_utils import process_vision_info
+from rich.console import Console
+from transformers import AutoProcessor, Qwen3VLMoeForConditionalGeneration
```
Comment (s1lent4gnt) [inline src/lerobot/data_processing/sarm_annotations/subtask_annotation.py:77]: The subtask annotation can only use Qwen models family, maybe it's better to make it work with other models that maybe better than Qwen in the future.

File: src/lerobot/policies/act/modeling_act.py [hunk 7]
```diff
@@ -50,6 +50,7 @@ class ACTPolicy(PreTrainedPolicy):
     def __init__(
         self,
         config: ACTConfig,
+        **kwargs,
```
Comment (s1lent4gnt) [inline src/lerobot/policies/act/modeling_act.py:53]: We need to add per-sample losses.

File: src/lerobot/policies/diffusion/modeling_diffusion.py [hunk 8]
```diff
@@ -56,6 +56,7 @@ class DiffusionPolicy(PreTrainedPolicy):
     def __init__(
         self,
         config: DiffusionConfig,
+        **kwargs,
```
Comment (s1lent4gnt) [inline src/lerobot/policies/diffusion/modeling_diffusion.py:59]: We need to add per-sample losses here too.

File: src/lerobot/policies/groot/modeling_groot.py [hunk 9]
```diff
@@ -49,7 +49,7 @@ class GrootPolicy(PreTrainedPolicy):
     name = "groot"
     config_class = GrootConfig
 
-    def __init__(self, config: GrootConfig):
+    def __init__(self, config: GrootConfig, **kwargs):
```
Comment (s1lent4gnt) [inline src/lerobot/policies/groot/modeling_groot.py:52]: We need to add per-sample losses here too.

File: src/lerobot/policies/tdmpc/modeling_tdmpc.py [hunk 10]
```diff
@@ -65,6 +65,7 @@ class TDMPCPolicy(PreTrainedPolicy):
     def __init__(
         self,
         config: TDMPCConfig,
+        **kwargs,
```
Comment (s1lent4gnt) [inline src/lerobot/policies/tdmpc/modeling_tdmpc.py:68]: We need to add per-sample losses here too

File: src/lerobot/policies/xvla/modeling_xvla.py [hunk 11]
```diff
@@ -273,7 +273,7 @@ class XVLAPolicy(PreTrainedPolicy):
     config_class = XVLAConfig
     name = "xvla"
 
-    def __init__(self, config: XVLAConfig):
+    def __init__(self, config: XVLAConfig, **kwargs):
```
Comment (s1lent4gnt) [inline src/lerobot/policies/xvla/modeling_xvla.py:276]: We need to add per-sample losses here too.

File: src/lerobot/policies/vqbet/modeling_vqbet.py [hunk 12]
```diff
@@ -47,6 +47,7 @@ class VQBeTPolicy(PreTrainedPolicy):
     def __init__(
         self,
         config: VQBeTConfig | None = None,
+        **kwargs,
```
Comment (s1lent4gnt) [inline src/lerobot/policies/vqbet/modeling_vqbet.py:50]: We need to add per-sample losses here too.

File: src/lerobot/policies/sarm/__init__.py [hunk 13]
```diff
@@ -0,0 +1,37 @@
...
+    make_sarm_pre_post_processors,
+)
+
+__all__ = [
+    "SARMConfig",
+    "SARMRewardModel",
+    "StageTransformer",
+    "SubtaskTransformer",
+    "gen_stage_emb",
+    "SARMEncodingProcessorStep",
```
Comment (s1lent4gnt) [inline src/lerobot/policies/sarm/__init__.py]: Do we need to add those here?

File: src/lerobot/policies/sarm/compute_rabc_weights.py [hunk 14]
```diff
@@ -0,0 +1,821 @@
...
+        --reward-model-path pepijn223/sarm_single_uni4 \\
+        --stride 5
+
+    # Visualize predictions only (no RABC computation)
+    python src/lerobot/policies/sarm/compute_rabc_weights.py \\
+        --dataset-repo-id lerobot/aloha_sim_insertion_human \\
+        --reward-model-path pepijn223/sarm_single_uni4 \\
+        --visualize-only \\
+        --num-visualizations 5
+
```
Comment (s1lent4gnt) [inline src/lerobot/policies/sarm/compute_rabc_weights.py]: Nice to change `RABC` to `RA-BC`

File: src/lerobot/policies/sarm/compute_rabc_weights.py [hunk 15]
```diff
@@ -0,0 +1,821 @@
...
+        img = np.transpose(img, (1, 2, 0))
+    if img.dtype != np.uint8:
+        # Handle normalized images (may have negative values or values > 1)
+        img = img.astype(np.float32)
+        img = (img - img.min()) / (img.max() - img.min() + 1e-8)  # Normalize to [0, 1]
+        img = (img * 255).astype(np.uint8)
+    return img
+
+
+def visualize_episode(
```
Comment (s1lent4gnt) [inline src/lerobot/policies/sarm/compute_rabc_weights.py:136]: Does this works the same as `visualize_episode` function in `subtask_annotation.py`

File: docs/source/sarm.mdx (hunk 1 above)
Comment (pkooij) [reply docs/source/sarm.mdx:229]: Yes it for easy copy paste, and its in a select box

File: src/lerobot/data_processing/sarm_annotations/subtask_annotation.py (hunk 5 above)
Comment (pkooij) [reply src/lerobot/data_processing/sarm_annotations/subtask_annotation.py:1]: Maybe not now, lets see how it goes with community?

File: src/lerobot/data_processing/sarm_annotations/subtask_annotation.py (hunk 6 above)
Comment (pkooij) [reply src/lerobot/data_processing/sarm_annotations/subtask_annotation.py:77]: Yes that's a next step on my todo list together with making this more universal for different annotations

File: src/lerobot/policies/sarm/__init__.py (hunk 13 above)
Comment (pkooij) [reply src/lerobot/policies/sarm/__init__.py]: No I don't think so indeed

File: src/lerobot/policies/sarm/compute_rabc_weights.py (hunk 15 above)
Comment (pkooij) [reply src/lerobot/policies/sarm/compute_rabc_weights.py:136]: No this is to visualize sarm progress, annotations only show the subtasks boundaries

File: src/lerobot/scripts/lerobot_train.py [hunk 16]
```diff
@@ -248,6 +275,29 @@ def train(cfg: TrainPipelineConfig, accelerator: Accelerator | None = None):
         logging.info("Creating optimizer and scheduler")
     optimizer, lr_scheduler = make_optimizer_and_scheduler(cfg, policy)
 
+    # Load precomputed SARM progress for RA-BC if enabled
+    # Generate progress using: examples/sarm/compute_rabc_weights.py
```
Comment (s1lent4gnt) [inline src/lerobot/scripts/lerobot_train.py]: This should be : src/lerobot/policies/sarm/compute_rabc_weights.py

**Review by CarolinePascal — DISMISSED** [review]
lgtm !

**Review by CarolinePascal — APPROVED** [review]
